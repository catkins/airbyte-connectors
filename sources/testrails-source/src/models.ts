import {DateTime} from 'luxon';

export interface TimeWindow {
  before?: DateTime;
  after?: DateTime;
}
