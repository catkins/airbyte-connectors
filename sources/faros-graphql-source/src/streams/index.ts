import {FarosGraph} from './faros-graph';

export {FarosGraph};
