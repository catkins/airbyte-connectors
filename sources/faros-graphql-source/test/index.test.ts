import {
  AirbyteLogger,
  AirbyteLogLevel,
  AirbyteSpec,
  SyncMode,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import * as sut from '../src/index';
import {GraphQLVersion} from '../src/index';

let graphExists = false;
const nodes = [{k1: 'v1'}, {k2: 'v2'}];
jest.mock('faros-js-client', () => {
  return {
    FarosClient: jest.fn().mockImplementation(() => {
      return {
        graphExists: () => {
          return graphExists;
        },
        nodeIterable: () => {
          return {
            async *[Symbol.asyncIterator](): AsyncIterator<any> {
              for (const item of nodes) {
                yield item;
              }
            },
          };
        },
      };
    }),
  };
});

describe('index', () => {
  const logger = new AirbyteLogger(
    // Shush messages in tests, unless in debug
    process.env.LOG_LEVEL === 'debug'
      ? AirbyteLogLevel.DEBUG
      : AirbyteLogLevel.FATAL
  );

  test('spec', async () => {
    const source = new sut.FarosGraphSource(logger);
    await expect(source.spec()).resolves.toStrictEqual(
      new AirbyteSpec(readResourceFile('spec.json'))
    );
  });

  test('check connection failure', async () => {
    const source = new sut.FarosGraphSource(logger);
    await expect(source.checkConnection({} as any)).resolves.toStrictEqual([
      false,
      new VError('Missing api_url'),
    ]);
    await expect(
      source.checkConnection({api_url: 'x'} as any)
    ).resolves.toStrictEqual([false, new VError('Missing api_key')]);
    await expect(
      source.checkConnection({api_url: 'x', api_key: 'y'} as any)
    ).resolves.toStrictEqual([false, new VError('Missing graphql_api')]);
    await expect(
      source.checkConnection({
        api_url: 'x',
        api_key: 'y',
        graphql_api: GraphQLVersion.V1,
      } as any)
    ).resolves.toStrictEqual([false, new VError('Missing graph')]);
    await expect(
      source.checkConnection({
        api_url: 'x',
        api_key: 'y',
        graphql_api: GraphQLVersion.V1,
        graph: 'default',
      } as any)
    ).resolves.toStrictEqual([false, new VError('Graph does not exist!')]);
  });

  test('check connection success', async () => {
    const source = new sut.FarosGraphSource(logger);
    graphExists = true;
    await expect(
      source.checkConnection({
        api_url: 'x',
        api_key: 'y',
        graphql_api: GraphQLVersion.V1,
        graph: 'default',
      } as any)
    ).resolves.toStrictEqual([true, undefined]);
  });

  test('full_refresh sync mode', async () => {
    const source = new sut.FarosGraphSource(logger);
    graphExists = true;
    const iter = source
      .streams({} as any)[0]
      .readRecords(SyncMode.FULL_REFRESH, undefined, undefined);

    const records = [];
    for await (const record of iter) {
      records.push(record);
    }

    expect(records).toMatchSnapshot();
  });
});
