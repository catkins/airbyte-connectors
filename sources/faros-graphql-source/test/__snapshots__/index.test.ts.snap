// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index full_refresh sync mode 1`] = `
Array [
  Object {
    "k1": "v1",
  },
  Object {
    "k2": "v2",
  },
]
`;

exports[`index incremental sync mode - V1 1`] = `
Array [
  Object {
    "k1": "v1",
    "metadata": Object {
      "refreshedAt": 12,
    },
  },
  Object {
    "k2": "v2",
    "metadata": Object {
      "refreshedAt": 23,
    },
  },
]
`;

exports[`index incremental sync mode - V2 1`] = `
Array [
  Object {
    "k1": "v1",
    "refreshedAt": "2022-10-19T22:02:14.483165+00:00",
  },
  Object {
    "k2": "v2",
    "refreshedAt": "2023-11-14T22:13:20.000Z",
  },
]
`;
